link-grammar/dict-common/dict-api.h

```diff
@@ -25,7 +25,9 @@ LINK_BEGIN_DECLS
  * the public API to the link-parser system.
  */
 
+#ifdef USE_VITERBI
 Dictionary dictionary_create_from_utf8(const char * input);
+#endif
```

Comment:
Some variant of this should almost certainly be in the public API.  Viz, a way to provide dictionary contents, without requiring files.  Not clear how affix and regex files could be handled, though... maybe more string routines.

Reply:
Maybe:
`Dictionary dictionary_create_from_string(const char *dict, const char *affix, const char *regex, const char *knowledge , const char *constituent_knowledge)`
when the one that are not used can be NULL.
The returned value should probably not be `const`, so API for manipulated the dict can be later added.

Reply:
Yeah, seems reasonable.   C still doesn't have default argument values, like C++, right?

Reply:
> C still doesn't have default argument values, like C++, right?

No such thing yet in C, but maybe it can be emulated by vararg or sentinel values, like:
``` c
# The rest assumed to be in files
dictionary_create_from_string(dictstr, NULL);
dictionary_create_from_string(dictstr, , affixstr, NULL);
# DICT_FROM_FILE means read from standard location
dictionary_create_from_string(dictstr, DICT_FROM_FILE, NULL, DICT_FROM_FILE, DICT_FROM_FILE);
```

Most probably the language should also be added - as first argument.

It can even be made more generic - to read from any source,.e.g. using URL-like prefixes like `file:` or `sqlite:` (default is string), and an optional additional argument can be a search path. In that case the function name can be just `dictionary_create()`.